File: leetcode/lc-us/cpp/1071.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
    static bool is_common_divisor(const string &str, const string &divisor) {
        for (int i = 0; i < str.size(); i += (int) divisor.size()) {
            if (str.substr(i, divisor.size()) != divisor) {
                return false;
            }
        }
        return true;
    }

public:
    string gcdOfStrings(string str1, string str2) {
        int n = (int) str1.size(), m = (int) str2.size();
        for (int len = __gcd(n, m); len >= 1; len--) {
            string divisor = str1.substr(0, len);
            if (n % len || m % len) continue;
            if (is_common_divisor(str1, divisor) && is_common_divisor(str2, divisor)) return divisor;
        }
        return "";
    }
};

File: leetcode/lc-us/py/1071.py
import math


class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        n, m = len(str1), len(str2)
        max_len = math.gcd(n, m)

        def valid(length: int) -> bool:
            for i in range(0, n - length + 1, length):
                if str1[i: i + length] != str1[0: length]:
                    return False
            for i in range(0, m - length + 1, length):
                if str2[i:i + length] != str1[0:length]:
                    return False
            return True

        for length in range(max_len, 0, -1):
            if n % length or m % length:
                continue
            if valid(length):
                return str1[:length]
        return ''
